bouncer/src/repo/lib/repo_license.cpp

```diff
@@ -0,0 +1,169 @@
+*  GNU Affero General Public License for more details.
+*
+*  You should have received a copy of the GNU Affero General Public License
+*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "repo_license.h"
+
+namespace Licensing
+{
+	// definitions for class static variables
+	repo::lib::RepoUUID LicenseValidator::instanceId;
+	std::string LicenseValidator::licenseStr;
+	std::unique_ptr<Cryptolens> LicenseValidator::cryptolensHandle;
+
+	void LicenseValidator::RunActivation()
+	{
+#ifdef REPO_LICENSE_CHECK
+		// only run once 
+		if (!instanceId.isDefaultValue() && cryptolensHandle) return;
+
+		// Setting up the handle
+		licenseStr = GetLicenseString();
+		cryptolens::Error e;
+		cryptolensHandle = std::make_unique<Cryptolens>(e);
+		cryptolensHandle->signature_verifier.set_modulus_base64(e, pubKeyModulus);
+		cryptolensHandle->signature_verifier.set_exponent_base64(e, pubKeyExponent);
+		// Using machine code to store the instance UUID instead, as per
+		// https://help.cryptolens.io/licensing-models/containers
+		instanceId = repo::lib::RepoUUID::createUUID();
+		cryptolensHandle->machine_code_computer.set_machine_code(e, instanceId.toString());
+
+		// License activation 
+		cryptolens::optional<cryptolens::LicenseKey> licenseKey =
+			cryptolensHandle->activate_floating
+			(
+				e, // Object used for reporting if an error occured
+				authToken, // Cryptolens Access Token
+				productId, // Product key associated with the 3drepo.io pipeline
+				licenseStr, // License Key
+				floatingTimeIntervalSec // The amount of time the user has to wait before an actived machine (or session in our case) is taken off the license.
+			);
+
+		// dealing with early bail out scenarios
+		repoInfo << "****License activation summary****";
+		if (e)
+		{
+			cryptolens::ActivateError error = cryptolens::ActivateError::from_reason(e.get_reason());
+			repoInfo << "- server message: " << error.what();
+			repoInfo << "- server respose ok: false";
+			repoInfo << "- session not added to license";
+			throw repo::lib::RepoInvalidLicenseException();
+		}
+		else if (!licenseKey)
+		{
+			repoInfo << "- server respose ok: false";
+			repoInfo << "- session not added to license. Error license LicenseKey is null";
+			throw repo::lib::RepoInvalidLicenseException();
+		}
+		// dealing with the license check logic once all info is present
+		else
+		{
+			// printing out the result
+			std::string notes = licenseKey->get_notes().has_value() ?
+				licenseKey->get_notes().value() : "";
+			int noUsedInsances = licenseKey->get_activated_machines().has_value() ?
+				licenseKey->get_activated_machines()->size() : -1;
+			int maxInstances = licenseKey->get_maxnoofmachines().has_value() ?
+				licenseKey->get_maxnoofmachines().value() : -1;
+			bool licenseBlocked = licenseKey->get_block();
+			bool licenseExpired = licenseKey->check().has_expired(time(0)) ? true : false;
```

Comment:
```suggestion
			bool licenseExpired = licenseKey->check().has_expired(time(0));
```

Reply:
this does not compile as their bool operator definition is marked explicit, but adding a static cast to a bool makes sure the returned class from `has_expired` correctly. I've done it that way to make it look neater.